osu.Server.QueueProcessor/QueueProcessor.cs

```diff
@@ -214,9 +227,21 @@ public virtual MySqlConnection GetDatabaseConnection()
         }
 
         /// <summary>
-        /// Implement to process a single item from the queue.
+        /// Implement to process a single item from the queue. Will only be invoked if <see cref="ProcessResults"/> is not implemented.
         /// </summary>
         /// <param name="item">The item to process.</param>
-        protected abstract void ProcessResult(T item);
+        protected virtual void ProcessResult(T item)
+        {
+        }
+
+        /// <summary>
+        /// Implement to process batches of items from the queue.
+        /// </summary>
+        /// <param name="items">The items to process.</param>
+        protected virtual void ProcessResults(IEnumerable<T> items)
+        {
+            foreach (var item in items)
+                ProcessResult(item);
```

Comment:
I'm a little bit worried about this... Suppose that item 0 passes, but an exception is thrown in item 1. The `ContinueWith()` above will retry both item 0 and item 1 due to matching the single exception. Is this intentional?

Reply:
This was intentional. If an exception is thrown, we don't know which items (if any) got processed, so they should be processed an extra time. For individual item failures, the new `Failed` property can be used.

Queue processor implementations to date have been written in a way where processing an already processed item is either a no-op, or safe via other means.

Reply:
Yeah I guess my expectation would just be for this to be implemented as a try-catch setting `Failed` to true, but if you say it's fine then so be it. `BatchSize` of 1 is probably going to be the most common among e.g. diffcalc/ppcalc processors.

Reply:
That won't work because the full items list needs to be sent to the implementation in one go to actually make batching useful.

Reply:
I meant just for this base implementation.